### PR TITLE
Support lateral view explode map

### PR DESCRIPTION
**Pull request by ljfgem**
Added the case of `aliasOperands.size() == 4` in `ParseTreeBuilder.visitLateralViewExplode` to support lateral view explode map.

Tests:
1. Unit tests
2. Test on affected view, which could be translated well with this patch
3. Integration test, no regression

### CODE REVIEW

File: coral-hive/src/main/java/com/linkedin/coral/hive/hive2rel/parsetree/ParseTreeBuilder.java [hunk 1]
```diff
@@ -222,38 +222,55 @@ private SqlNode visitLateralViewInternal(ASTNode node, ParseContext ctx, boolean
 
   private SqlNode visitLateralViewExplode(List<SqlNode> sqlNodes, List<SqlNode> aliasOperands,
       SqlCall tableFunctionCall, boolean isOuter) {
-    checkState(aliasOperands.size() == 3);
+    final int operandCount = aliasOperands.size();
+    // array if operandCount == 3, map if operandCount == 4
```
Comment (shenodaguirguis) [inline coral-hive/src/main/java/com/linkedin/coral/hive/hive2rel/parsetree/ParseTreeBuilder.java:226]: suggestion: would be nice to add example syntax for array and map, to be clear why operandCount == 3 for array and 4 for map. Something like:

`map if operandCount == 4:  LATERAL VIEW EXPLODE(op0) op1 AS op2, op3 `

File: coral-hive/src/main/java/com/linkedin/coral/hive/hive2rel/parsetree/ParseTreeBuilder.java [hunk 2]
```diff
@@ -222,38 +222,55 @@ private SqlNode visitLateralViewInternal(ASTNode node, ParseContext ctx, boolean
...
+    checkState(operandCount == 3 || operandCount == 4,
+        format("Unsupported LATERAL VIEW EXPLODE operand number: %d", operandCount));
     // TODO The code below assumes LATERAL VIEW is used with UNNEST EXPLODE only. It should be made more generic.
```
Comment (shenodaguirguis) [inline coral-hive/src/main/java/com/linkedin/coral/hive/hive2rel/parsetree/ParseTreeBuilder.java:229]: do you know if there is an issue created for this? If not, do you mind creating one and reference it here? (and please create a parallel internal BDP jira in the backlog)

File: coral-hive/src/main/java/com/linkedin/coral/hive/hive2rel/parsetree/ParseTreeBuilder.java [hunk 3]
```diff
@@ -222,38 +222,55 @@ private SqlNode visitLateralViewInternal(ASTNode node, ParseContext ctx, boolean
...
+          aliasCall/*lateralCall*/, JoinConditionType.NONE.symbol(ZERO), null);
+    } else { // unnest explode map
+      unnestCall = HiveExplodeOperator.EXPLODE.createCall(ZERO, unnestOperand);
+      SqlNode rightSelect = new SqlSelect(ZERO, null, new SqlNodeList(ImmutableList.of(SqlIdentifier.star(ZERO)), ZERO),
+          unnestCall, null, null, null, null, null, null, null);
+      SqlNode lateralCall = SqlStdOperatorTable.LATERAL.createCall(ZERO, rightSelect);
+      SqlCall aliasCall = SqlStdOperatorTable.AS.createCall(ZERO, lateralCall, aliasOperands.get(1),
+          aliasOperands.get(2), aliasOperands.get(3));
+      return new SqlJoin(ZERO, sqlNodes.get(1), SqlLiteral.createBoolean(false, ZERO), JoinType.COMMA.symbol(ZERO),
+          aliasCall/*lateralCall*/, JoinConditionType.NONE.symbol(ZERO), null);
```
Comment (shenodaguirguis) [inline coral-hive/src/main/java/com/linkedin/coral/hive/hive2rel/parsetree/ParseTreeBuilder.java:272]: this block is identical to block lines 257-263, can we combine them after the if statement?

File: coral-schema/src/test/java/com/linkedin/coral/schema/avro/ViewToAvroSchemaConverterTests.java [hunk 4]
```diff
@@ -406,17 +406,17 @@ public void testLateralViewArrayWithComplexType() {
...
     String viewSql = "CREATE VIEW v AS " + "SELECT bl.Id AS Id_View_Col, t.Col1, t.Col2 " + "FROM baselateralview bl "
         + "LATERAL VIEW explode(bl.Map_Col_String) t as Col1, Col2";
 
     TestUtils.executeCreateViewQuery("default", "v", viewSql);
 
     ViewToAvroSchemaConverter viewToAvroSchemaConverter = ViewToAvroSchemaConverter.create(hiveMetastoreClient);
-    viewToAvroSchemaConverter.toAvroSchema("default", "v");
+    Schema actualSchema = viewToAvroSchemaConverter.toAvroSchema("default", "v");
+
+    Assert.assertEquals(actualSchema.toString(true), TestUtils.loadSchema("testLateralViewMap-expected.avsc"));
```
Comment (shenodaguirguis) [inline coral-schema/src/test/java/com/linkedin/coral/schema/avro/ViewToAvroSchemaConverterTests.java:419]: nit: assertEquals(expected, actual)
here & elsewhere in your diff

**Review by shenodaguirguis — CHANGES_REQUESTED** [review]
thanks for the extensive testing @ljfgem !
left couple minor comments, LGTM o.w.

File: coral-hive/src/main/java/com/linkedin/coral/hive/hive2rel/parsetree/ParseTreeBuilder.java (hunk 1 above)
Comment (ljfgem) [reply coral-hive/src/main/java/com/linkedin/coral/hive/hive2rel/parsetree/ParseTreeBuilder.java]: Thanks for suggestion, added.

File: coral-hive/src/main/java/com/linkedin/coral/hive/hive2rel/parsetree/ParseTreeBuilder.java (hunk 2 above)
Comment (ljfgem) [reply coral-hive/src/main/java/com/linkedin/coral/hive/hive2rel/parsetree/ParseTreeBuilder.java:230]: Opened an issue (https://github.com/linkedin/coral/issues/108) and created a ticket.

File: coral-schema/src/test/java/com/linkedin/coral/schema/avro/ViewToAvroSchemaConverterTests.java (hunk 4 above)
Comment (ljfgem) [reply coral-schema/src/test/java/com/linkedin/coral/schema/avro/ViewToAvroSchemaConverterTests.java:419]: The definition of `assertEquals` is:
```
  static public void assertEquals(String actual, String expected) {
    assertEquals(actual, expected, null);
  }
```
And we used `assertEquals(String actual, String expected)` for all the tests in coral

Comment (shenodaguirguis) [reply coral-schema/src/test/java/com/linkedin/coral/schema/avro/ViewToAvroSchemaConverterTests.java:419]: http://junit.sourceforge.net/javadoc/org/junit/Assert.html#assertEquals(long,%20long) ?

Comment (shenodaguirguis) [reply coral-schema/src/test/java/com/linkedin/coral/schema/avro/ViewToAvroSchemaConverterTests.java:419]: oh, you are using testng, not junit :-) sorry

**Review by shenodaguirguis — APPROVED** [review]

**Pull request merged**